lib/AST/ASTPrinter.cpp

```diff
@@ -5180,46 +5334,82 @@ void PrintAST::visitMemberRefExpr(MemberRefExpr *expr) {
 }
 
 void PrintAST::visitSubscriptExpr(SubscriptExpr *expr) {
+  visit(expr->getBase());
+  Printer << "[";
+  printArgument(expr->getArgs()->front());
+  Printer << "]";
 }
 
 void PrintAST::visitEnumIsCaseExpr(EnumIsCaseExpr *expr) {
+  visit(expr);
+  Printer << " is ";
+  printTypeLoc(expr->getCaseTypeRepr());
 }
 
 void PrintAST::visitForceValueExpr(ForceValueExpr *expr) {
+  visit(expr->getSubExpr());
+  Printer << "!";
 }
 
 void PrintAST::visitKeyPathDotExpr(KeyPathDotExpr *expr) {
+  
 }
 
 void PrintAST::visitAutoClosureExpr(AutoClosureExpr *expr) {
   visit(expr->getSingleExpressionBody());
 }
 
 void PrintAST::visitCaptureListExpr(CaptureListExpr *expr) {
+  auto captureList = expr->getCaptureList();
```

Comment:
`AbstractClosureExpr` is the base class that's inherited by both `ClosureExpr` (for explicitly written closures in the source), and for `AutoClosureExpr` (for implicitly created closures). We need to be able to handle both because both can have capture lists, and so `captureList->getClosureBody()` returns an `AbstractClosureExpr`. IIRC the main case where we introduce capture lists for auto-closures is for key-paths as functions. We may want to try detect that case and print using the key-path syntax instead, but I think for now the closure syntax is fine.